### PR TITLE
expression: handle max_allowed_packet warnings for from_base64 function.

### PR DESCRIPTION
**Pull request by supernan1994**
### What problem does this PR solve? 
Return NULL and a warning when the result of from_base64 exceeds max_allowed_packetbs.
to #7153 

### What is changed and how it works?
Before this PR:
```
mysql> show variables like 'max_allowed_packet';
+--------------------+-------+
| Variable_name      | Value |
+--------------------+-------+
| max_allowed_packet | 2     |
+--------------------+-------+
1 row in set (0.01 sec)

mysql> select from_base64("YWJj");
+---------------------+
| from_base64("YWJj") |
+---------------------+
| abc                 |
+---------------------+
1 row in set (0.04 sec)
```
After this PR:
```
mysql> show variables like 'max_allowed_packet';
+--------------------+-------+
| Variable_name      | Value |
+--------------------+-------+
| max_allowed_packet | 2     |
+--------------------+-------+
1 row in set (0.01 sec)

mysql> select from_base64("YWJj");
+---------------------+
| from_base64("YWJj") |
+---------------------+
| NULL                |
+---------------------+
1 row in set, 1 warning (0.00 sec)

mysql> show warnings;
+---------+------+----------------------------------------------------------------------------+
| Level   | Code | Message                                                                    |
+---------+------+----------------------------------------------------------------------------+
| Warning | 1301 | Result of from_base64() was larger than max_allowed_packet (2) - truncated |
+---------+------+----------------------------------------------------------------------------+
1 row in set (0.00 sec)
```
This PR adds `base64NeededDecodedLength` function. This function calculates the length of `from_base64`'s result before executing core of `from_base64` logic, and compares the length with global variable `max_allowed_packet`.

### Check List 

Tests 

 - Unit test

Related changes

 - Need to be included in the release note
please note:
```
Return `NULL` when the result of function `FROM_BASE64` is larger than `max_allowed_packet`
```

### CODE REVIEW

**Comment by supernan1994** [comment]
@zz-jason PTAL

**Comment by shenli** [comment]
@supernan1994 Thanks!

File: expression/builtin_string.go [hunk 1]
```diff
@@ -2959,18 +2959,42 @@ func (c *fromBase64FunctionClass) getFunction(ctx sessionctx.Context, args []Exp
...
-	sig := &builtinFromBase64Sig{bf}
+	sig := &builtinFromBase64Sig{bf, maxAllowedPacket}
 	return sig, nil
 }
 
+// base64NeededDecodedLength return the base64 decoded string length.
+func base64NeededDecodedLength(n int) int {
+	// Returns -1 indicate the result will overflow.
+	if strconv.IntSize == 64 {
+		if n > 3074457345618258602 {
```
Comment (zz-jason) [inline expression/builtin_string.go]: Is this value equal to `math.MaxInt64`?

File: expression/builtin_string.go [hunk 2]
```diff
@@ -2959,18 +2959,42 @@ func (c *fromBase64FunctionClass) getFunction(ctx sessionctx.Context, args []Exp
...
+			return -1
+		}
+	} else {
+		if n > 715827882 {
+			return -1
+		}
+	}
+	return n * 3 / 4
```
Comment (zz-jason) [inline expression/builtin_string.go]: How about `n/4*3`, then the above validation can be removed?

File: expression/builtin_string.go [hunk 3]
```diff
@@ -2959,18 +2959,42 @@ func (c *fromBase64FunctionClass) getFunction(ctx sessionctx.Context, args []Exp
...
 
+// base64NeededDecodedLength return the base64 decoded string length.
+func base64NeededDecodedLength(n int) int {
+	// Returns -1 indicate the result will overflow.
+	if strconv.IntSize == 64 {
+		if n > 3074457345618258602 {
+			return -1
+		}
+	} else {
+		if n > 715827882 {
```
Comment (coocood) [inline expression/builtin_string.go]: How did you get this number?

File: expression/builtin_string.go (hunk 1 above)
Comment (supernan1994) [reply expression/builtin_string.go]: No, it's equals to `0x2AAAAAAAAAAAAAAALL`, I copy it from mysql.
https://github.com/mysql/mysql-server/blob/5.7/mysys/base64.c#L69
I haven't figured out how it works, do you have any ideas?

File: expression/builtin_string.go (hunk 3 above)
Comment (supernan1994) [reply expression/builtin_string.go]: I copy it from mysql.
https://github.com/mysql/mysql-server/blob/5.7/mysys/base64.c#L69
I haven't figured out how it works, do you have any ideas?

File: expression/builtin_string.go (hunk 1 above)
Comment (shenli) [reply expression/builtin_string.go]: It is better to define a constant instead of using a magic number.

File: expression/builtin_string.go (hunk 2 above)
Comment (supernan1994) [reply expression/builtin_string.go]: when `n%4 > 4/3` , `n/4*3` is not equal to `n*3/4`.
mysql calculates decode length by `n*3/4`, I think we'd better use the same formula.

File: expression/builtin_string.go (hunk 1 above)
Comment (zz-jason) [reply expression/builtin_string.go]: It is `math.MaxInt64/3`

File: expression/builtin_string.go (hunk 3 above)
Comment (zz-jason) [reply expression/builtin_string.go]: It is `math.MaxInt32/3`

File: expression/builtin_string.go (hunk 1 above)
Comment (zz-jason) [reply expression/builtin_string.go]: how about:
```go
if strconv.IntSize == 64 && n > math.MaxInt64/3 {
	return -1
}
if strconv.IntSize == 32 && n > math.MaxInt32/3 {
	return -1
}
```

File: expression/builtin_string.go (hunk 2 above)
Comment (zz-jason) [reply expression/builtin_string.go]: ok

File: expression/builtin_string.go (hunk 1 above)
Comment (supernan1994) [reply expression/builtin_string.go]: Sure, Thanks a lot!! I will correct it.
I still confuse about why it is `math.MaxInt64/3`. Does that mean `base64_decode` function malloc 3 times of source string length during the process? It doesn't make sense.

Comment (zz-jason) [reply expression/builtin_string.go]: This check is to guarantee that the later `n*3` instruction, which is part of `n*3/4`, does overflow the max int64 or int32 value.

Comment (supernan1994) [reply expression/builtin_string.go]: Oh, I get it, thanks~

Comment (zz-jason) [reply expression/builtin_string.go]: 😄

**Comment by supernan1994** [comment]
@zz-jason DONE. PTAL

File: expression/builtin_string_test.go [hunk 4]
```diff
@@ -1609,6 +1609,62 @@ func (s *testEvaluatorSuite) TestFromBase64(c *C) {
...
+
+	for _, test := range tests {
+		resultType := &types.FieldType{Tp: mysql.TypeVarchar, Flen: mysql.MaxBlobWidth}
+		base := baseBuiltinFunc{args: args, ctx: s.ctx, tp: resultType}
+		fromBase64 := &builtinFromBase64Sig{base, test.maxAllowPacket}
+
+		input := chunk.NewChunkWithCapacity(colTypes, 1)
+		input.AppendString(0, test.args)
+		res, isNull, err := fromBase64.evalString(input.GetRow(0))
+		c.Assert(err, IsNil)
```
Comment (crazycs520) [inline expression/builtin_string_test.go]: add `c.Assert(isNull, Equals, test.isNil)`
then we can remove line #1653 and else branch in line #1662

**Comment by zz-jason** [comment]
@supernan1994 It would be great if you can cherrypick this bugfix to the `release-2.0` branch 😸

**Comment by supernan1994** [comment]
@zz-jason OK!

**Pull request merged**